homeassistant/components/zwave_js/migrate.py

```diff
@@ -1,27 +1,354 @@
+    zwave_data: dict[str, ZWaveMigrationData],
+    zwave_js_data: dict[str, ZWaveJSMigrationData],
+) -> LegacyZWaveMappedData:
+    """Map Z-Wave node values onto Z-Wave JS node values."""
+    migration_map = LegacyZWaveMappedData()
+    zwave_proc_data: dict[
+        tuple[int, int, int, str, str | None, str | None],
+        ZWaveMigrationData,
+    ] = {}
+    zwave_js_proc_data: dict[
+        tuple[int, int, int, str, str | None, str | None],
+        ZWaveJSMigrationData,
+    ] = {}
+
+    for zwave_item in zwave_data.values():
+        zwave_js_property_name = CC_ID_LABEL_TO_PROPERTY.get(
+            zwave_item["command_class"], {}
+        ).get(zwave_item["command_class_label"])
+        item_id = (
+            zwave_item["node_id"],
+            zwave_item["command_class"],
+            zwave_item["node_instance"] - 1,
+            zwave_item["domain"],
+            zwave_item["unit_of_measurement"],
+            zwave_js_property_name,
+        )
+
+        # Remove duplicates that are not resolvable.
+        if item_id in zwave_proc_data:
+            zwave_proc_data.pop(item_id)
+            continue
+
+        zwave_proc_data[item_id] = zwave_item
+
+    for zwave_js_item in zwave_js_data.values():
+        # Only identify with property name if there is a command class label map.
+        if zwave_js_item["command_class"] in CC_ID_LABEL_TO_PROPERTY:
+            zwave_js_property_name = zwave_js_item["value_property_name"]
+        else:
+            zwave_js_property_name = None
+        item_id = (
+            zwave_js_item["node_id"],
+            zwave_js_item["command_class"],
+            zwave_js_item["endpoint_index"],
+            zwave_js_item["domain"],
+            zwave_js_item["unit_of_measurement"],
+            zwave_js_property_name,
+        )
+
+        # Remove duplicates that are not resolvable.
+        if item_id in zwave_js_proc_data:
+            zwave_js_proc_data.pop(item_id)
+            continue
+
+        zwave_js_proc_data[item_id] = zwave_js_item
+
+    for item_id, zwave_entry in zwave_proc_data.items():
+        zwave_js_entry = zwave_js_proc_data.pop(item_id, None)
+
+        if zwave_js_entry is None:
+            continue
+
+        migration_map.entity_entries[zwave_js_entry["entity_id"]] = zwave_entry
+        migration_map.device_entries[zwave_js_entry["device_id"]] = zwave_entry[
+            "device_id"
+        ]
+
+    return migration_map
+
+
+async def async_migrate_legacy_zwave(
+    hass: HomeAssistant,
+    zwave_config_entry: ConfigEntry,
+    zwave_js_config_entry: ConfigEntry,
+    migration_map: LegacyZWaveMappedData,
+) -> None:
+    """Perform Z-Wave to Z-Wave JS migration."""
+    dev_reg = async_get_device_registry(hass)
+    for zwave_js_device_id, zwave_device_id in migration_map.device_entries.items():
+        zwave_device_entry = dev_reg.async_get(zwave_device_id)
+        if not zwave_device_entry:
+            continue
+        dev_reg.async_update_device(
```

Comment:
Link device entries instead of updating area and name:
https://github.com/home-assistant/core/pull/39081#discussion_r520175152

Reply:
this is fine. We can't merge them at this point because we don't have merge logic in device registry.